File: LC 902.cpp
class Solution {
public:
    int atMostNGivenDigitSet(vector<string>& digits, int n) {
        auto s = to_string(n);
        int m = s.length(), dp[m];
        memset(dp, -1, sizeof(dp));
        function<int(int, bool, bool)> f = [&] (int i, bool is_limit, bool is_num) -> int {
            if (i >= m) return is_num;
            if (!is_limit && is_num && dp[i] >= 0) return dp[i];
            int res = 0;
            if (!is_num) {
                res = f(i + 1, false, false);
            }
            int up = is_limit ? s[i] - '0' : 9;
            for (const auto& t : digits) {
                int x = atoi(t.c_str());
                if (x <= up) {
                    res += f(i + 1, is_limit && x == up, true);
                } else {
                    break;
                }
            }
            if (!is_limit && is_num) {
                dp[i] = res;
            }
            return res;
        };
        return f(0, true, false);
    }
};

File: 1000/LC 902.go
func atMostNGivenDigitSet(digits []string, n int) int {
    s := strconv.Itoa(n)
    m := len(s)

    memo := make([]int, m)
    for i := range memo {
        memo[i] = -1
    }

    var dfs func(int, bool, bool) int
    dfs = func(i int, isLimit, isNum bool) (res int) {
        if i == m {
            if isNum {
                return 1
            }
            return
        }

        if !isLimit && isNum {
            p := &memo[i]
            if *p != -1 {
                return *p
            }
            defer func(){*p = res}()
        }

        if !isNum {
            res += dfs(i + 1, false, false)
        }

        up := byte('9')
        if isLimit {
            up = s[i]
        }
        for _, v := range digits {
            if v[0] <= up {
                res += dfs(i + 1, isLimit && v[0] == up, true)
            }
        }

        return
    }

    return dfs(0, true, false)
}